src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs

```diff
@@ -0,0 +1,196 @@
+        public void OnlyOneShiftMayBeActive()
+        {
+            int count = 0;
+            foreach (var @event in _events)
+            {
+                if (@event.StartsWith("ShiftStarted") && ++count > 1)
+                    Assert.Fail(DumpEvents("Shift started while another shift was active"));
+
+                if (@event.StartsWith("ShiftFinished"))
+                    --count;               
+            }
+        }
+
+        [Test]
+        public void AllTestsAreWithinShifts()
+        {
+            bool inShift = false;
+            foreach (var @event in _events)
+            {
+                if (@event.StartsWith("ShiftStarted"))
+                    inShift = true;
+                else if (@event.StartsWith("ShiftFinished"))
+                    inShift = false;
+                else
+                    if(!inShift)
+                        Assert.Fail(DumpEvents("Test event received outside of any shift"));
+            }
+        }
+
+        [Test]
+        public void ListEvents()
+        {
+            Console.WriteLine(DumpEvents("Events Received:"));
+        }
+        #region Test Data
+
+        static IEnumerable<TestFixtureData> GetParallelSuites()
+        {
+            yield return new TestFixtureData(
+                TestBuilder.MakeSuite("NUnit.TestData.ParallelExecutionData")
+                    .Containing(TestBuilder.MakeFixture(typeof(TestSetUpFixture))
+                        .Containing(typeof(TestFixture1), typeof(TestFixture2), typeof(TestFixture3))),
+                3);
+
+            yield return new TestFixtureData(
+                TestBuilder.MakeSuite("NUnit")
+                    .Containing(TestBuilder.MakeSuite("TestData")
+                        .Containing(TestBuilder.MakeSuite("ParallelExecutionData")
+                            .Containing(TestBuilder.MakeFixture(typeof(TestSetUpFixture))
+                                .Containing(typeof(TestFixture1), typeof(TestFixture2), typeof(TestFixture3))))),
+                3);
+        }
+
+        #endregion
+
+        #region ITestListener implementation
+
+        public void TestStarted(ITest test)
+        {
+            AddEvent("TestStarted " + test.FullName);
+        }
+
+        public void TestFinished(ITestResult result)
+        {
+            AddEvent("TestFinished " + result.FullName + " " + result.ResultState);
+        }
+
+        public void TestOutput(TestOutput output)
+        {
+            AddEvent(output.Stream + " " + output.Text.TrimEnd('\r', '\n'));
+        }
+
+        #endregion
+
+        #region Helper Methods
+
+        private void AddEvent(string @event)
+        {
+            lock (_events)
+                _events.Add(@event);
+        }
+
+        private string DumpEvents(string message)
+        {
+            var sb = new StringBuilder(message + NL);
+            foreach (string @event in _events)
+            {
+                sb.Append(@event);
+                sb.Append(NL);
```

Comment:
`new StringBuilder(message).AppendLine();`, and `sb.AppendLine(@event)` here? This would allow you to remove the `NL` declaration too.